From 6efd76f74e201df4e6b024c2a335d3aa9d2fb801 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Tue, 12 Aug 2025 14:25:18 +0200
Subject: [PATCH] HACK: kernelbase: Return parent-process as 'steam.exe' with
 WINE_ENABLE_STEAM_STUB

For games whose behavior depends on Proton's stub steam.exe, might come useful.
---
 dlls/kernelbase/debug.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/dlls/kernelbase/debug.c b/dlls/kernelbase/debug.c
index 9d9b4b36b41..d541892487b 100644
--- a/dlls/kernelbase/debug.c
+++ b/dlls/kernelbase/debug.c
@@ -1446,6 +1446,9 @@ DWORD WINAPI DECLSPEC_HOTPATCH GetModuleFileNameExA( HANDLE process, HMODULE mod
 {
     WCHAR *ptr;
     DWORD len;
+    static int count = 0;
+    char env_buf[16];
+    DWORD env_len = GetEnvironmentVariableA("WINE_ENABLE_STEAM_STUB", env_buf, sizeof(env_buf));
 
     TRACE( "(process=%p, module=%p, %p, %ld)\n", process, module, name, size );
 
@@ -1454,6 +1457,20 @@ DWORD WINAPI DECLSPEC_HOTPATCH GetModuleFileNameExA( HANDLE process, HMODULE mod
         SetLastError( ERROR_INVALID_PARAMETER );
         return 0;
     }
+
+    if (!count && env_len > 0 && atoi(env_buf) == 1
+        && process != GetCurrentProcess() && module == NULL)
+    {
+        const char* steam_path = "c:\\windows\\system32\\steam.exe";
+
+        len = strlen(steam_path);
+        lstrcpynA(name, steam_path, size);
+        count = 1;
+
+        FIXME("Process detected, Steam stub enabled..\n");
+        return len;
+    }
+
     if (process == GetCurrentProcess())
     {
         len = GetModuleFileNameA( module, name, size );
-- 
2.50.1

